guides/10_other-tutorials/building-an-mcp-client-with-gradio.md

```diff
@@ -0,0 +1,489 @@
+
+## What is MCP?
+
+The Model Context Protocol (MCP) standardizes how applications provide context to LLMs. It allows Claude to interact with external tools, like image generators, file systems, or APIs, etc.
+
+## Prerequisites
+
+- Python 3.10+
+- An Anthropic API key
+- Basic understanding of Python programming
+
+## Setup
+
+First, install the required packages:
+
+```bash
+pip install gradio anthropic mcp
+```
+
+Create a `.env` file in your project directory and add your Anthropic API key:
+
+```
+ANTHROPIC_API_KEY=your_api_key_here
+```
+
+## Part 1: Building the MCP Server
+
+The server provides tools that Claude can use. In this example, we'll create a server that generates images through [a HuggingFace space](https://huggingface.co/spaces/ysharma/SanaSprint).
+
+Create a file named `gradio_mcp_server.py`:
+
+```python
+from mcp.server.fastmcp import FastMCP
+import requests
+import json
+import sys
+import io
+import time
+
+# Set UTF-8 encoding for stdout/stderr
+sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')
+sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8', errors='replace')
+
+# Initialize FastMCP server
+mcp = FastMCP("huggingface_spaces_image_display")
+
+@mcp.tool()
+async def generate_image(prompt: str, width: int = 512, height: int = 512) -> str:
+    """Generate an image using SanaSprint model.
+    
+    Args:
+        prompt: Text prompt describing the image to generate
+        width: Image width (default: 512)
+        height: Image height (default: 512)
+    """
+    # API endpoint
+    base_url = "https://ysharma-sanasprint.hf.space/gradio_api/call/infer"
+    
+    # Parameters
+    payload = {
+        "data": [
+            prompt,               # User prompt
+            "0.6B",               # Smallest model
+            0,                    # Seed
+            True,                 # Randomize seed
+            width,                # Width
+            height,               # Height
+            4.0,                  # Guidance scale
+            2                     # Minimum steps
+        ]
+    }
+    
+    headers = {"Content-Type": "application/json"}
+    session = requests.Session()
+    
+    # Step 1: Make the POST request to get event ID
+    response = session.post(
+        base_url, 
+        headers=headers, 
+        json=payload, 
+        timeout=60
+    )
+    response_data = response.json()
+    
+    if "event_id" not in response_data:
+        return json.dumps({
+            "type": "error",
+            "message": "Failed to get event ID from service"
+        })
+    
+    event_id = response_data["event_id"]
+    
+    # Step 2: Poll for results
+    result_url = f"{base_url}/{event_id}"
+    
+    # Poll with timeout
+    start_time = time.time()
+    timeout = 60
+    
+    while (time.time() - start_time) < timeout:
```

Comment:
Not sure if intended but this entire loop can take longer than `timeout` seconds. The `iter_lines` loop can take an arbitrary amount of time.